### PR TITLE
feat: leverage IPFS node provided by Brave

### LINKED ISSUE #947: Detect and manage embedded go-ipfs in Brave
> This supersedes embedded js-ipfs (#716)

Brave Nightly will ship embedded go-ipfs soon and will expose some API endpoints for ipfs-companion to facilitate end-to-end integration. Custom WebExtension APIs will enable ipfs-companion to:
- read user preference on how to load IPFS resources (public gw, local gw, embedded one, disabled)
- read `$IPFS_PATH/config`  of embedded go-ipfs node managed by Brave
- start/stop the node

We should at the very minimum:

- add embedded go-ipfs to the list, if user enabled it in brave
- on initial run, use embedded node as a fallback
  - detect when embedded go-ipfs running, and save it as the default instead of 5001/8080 one
  - this should be fallback, we dont want to break Brave users who already use ipfs-companion+ipfs-desktop

TBD:

- start (and stop?) embedded node when user switches Node Type in Companion's Preferences
- hide/migrate embedded js-ipfs to embedded go-ipfs

### Details 

Excerpt from https://github.com/brave/brave-browser/issues/10220:

```
// Checks if a feature flag is enabled
chrome.ipfs.getIPFSEnabled((enabled: boolean) => {})

// Obtains a string representation of the resolve method
// method is one of the following strings:
// "ask" uses a gateway but also prompts them to install a local node
// "gateway" uses a gateway but also prompts them to install a local node
// "local" uses a gateway but also prompts them to install a local node
// "disabled" uses a gateway but also prompts them to install a local node
// Throws an error if IPFS feature flag is not enabled
chrome.ipfs.getResolveMethodType((method: string) => {})

// Launches a daemon
// Throws an error if IPFS feature flag is not enabled
// Throws an error if a local node is not installed
chrome.ipfs.launch((success: boolean) => {})

// Shutsdown a daemon
// Throws an error if IPFS feature flag is not enabled
// Throws an error if a local node is not installed
chrome.ipfs.shutdown((success: boolean) => {})

// Obtains the config contents of the local IPFS node
// Throws an error if IPFS feature flag is not enabled
// Throws an error if a local node is not installed
chrome.ipfs.getConfig((success: boolean, config: string) => {})

// Checks if the local node is installed
chrome.ipfs.getExecutableAvailable((available: boolean) => {})
```

#### `chrome.ipfs.*` behavior in Brave Nightly @ v1.20.36

- `getIPFSEnabled`  is enabled by default in Nightly, but we should check it as a formality
- `getResolveMethodType` is the actual hint of user's preference
  - `ask` is the default on an empty profile
  - `local`  – after "Enable IPFS" is clicked and go-ipfs daemon got installed 
- `launch`  
  - **the weird one.** returns `false` but also triggers install in the background, when claled later (after install is done) it starts node and returns `true`, but does not change `resolveMethodType` (it is still `ask`)
     - filled  https://github.com/brave/brave-browser/issues/13237 to clarify if it is a feature or a bug 
- `shutdown` always returns `true` (even if node was already off)
- `getExecutableAvailableget` returns true after install

### References

-  Native IPFS support in Brave: https://github.com/brave/brave-browser/issues/10220

### PR DESCRIPTION
**Pull request by lidel**
This PR adds support for IPFS node provided by Brave Nightly (https://github.com/brave/brave-browser/issues/10220) and closes #947

## Important!
https://github.com/ipfs/ipfs-docs/pull/581 needs merging once this work is completed so docs are up to date.

## TLDR

- detects `chrome.ipfs` APIs and adds "Provided by Brave" node type in Preferences
- ensures correct ports are used, even if user manually changes the config of Brave's node
- trigger "Enable IPFS" flow if user did not choose how to resolve `ipfs://` yet
- wait for Brave to download and install binary (if missing)
- this PR does not remove js-ipfs+chrome.sockets node type yet (but we update UI to indicate it is deprecated)

## Onboarding flow (IPFS already enabled in Brave)

When Companion is installed, user is presented  with Welcome screen that has "Brave user" section at the very top:

> onboarding-2021-01-04--14-02-49

Clicking on "Open Preferences" lets user to choose backend. In brave, "Provided by Brave" is added to the list:

> 

## Onboarding flow (IPFS not enabled yet)

The very first time (on an empty profile) is the same as before, but when "Provided by Brave"  is selected Companion checks if Brave user selected how to resolve IPFS resources.
If they did not make the choice yet (resolve method being `ask`) we open `ipfs://` URI in a new tab to trigger the decision via dropbar UI:

> 

Companion needs user to click "Enable IPFS"  – it will download go-ipfs binary and set local node as the default resolver of `ipfs://` and `ipns://` URIs.

For now, the triggers is just an image, but a dedicated HTML page with some text asking user to click "Enable IPFS" would be better, as downloading go-ipfs binary may take time and user won't see any visual feedback for a while.

Given available APIs, our extension is able to update the tab to indicate a different activation stage and provide some visual feedback:
1. Asking user to click on "Enable IPFS" button (**User action needed**)
2. Waiting for go-ipfs binary download
3. Waiting for `ipfs init`
4. Waiting for `ipfs daemon` to start

@jessicaschilling I think a simple static page(s) similar to https://github.com/ipfs-shipyard/ipfs-companion/pull/911 would do – are you able to come up with user-friendly text/prompt for the activation steps? First one is the most important, remaining ones are just progress updates.

### CODE REVIEW

**Comment by jessicaschilling** [comment]
Note: We'll also need to update https://docs.ipfs.io/how-to/companion-node-types accordingly.

**Comment by jessicaschilling** [comment]
@lidel Sending you a page under separate cover, since I assume we want to just put this on a gateway eventually. Let me know.

**Comment by jessicaschilling** [comment]
@lidel -- I've added https://github.com/ipfs/ipfs-docs/pull/581 for docs changes. Can you please make any edits you'd like, and merge when the rest of this work is complete? Thank you!

**Comment by lidel** [comment]
@jessicaschilling Thanks! I replaced activation trigger to use your onboarding page + trimmed one label (because its displayed only in Brave). 
I'm going to ship this to Beta, so we can have more eyes on this.
